### PR TITLE
ESQL: Added BytesRef offset randomization to function tests

### PR DESCRIPTION
**Pull request by ivancea**
Extracted from https://github.com/elastic/elasticsearch/pull/109205

Added a method to automatically randomize BytesRef parameters in ESQL function tests.

Will be merged after the original PR, which fixes an otherwise failing test.

### CODE REVIEW

File: x-pack/plugin/esql/src/test/java/org/elasticsearch/xpack/esql/expression/function/AbstractFunctionTestCase.java [hunk 1]
```diff
@@ -642,6 +642,46 @@ public interface ExpectedEvaluatorToString {
...
+     */
+    protected static List<TestCaseSupplier> randomizeBytesRefsOffset(List<TestCaseSupplier> testCaseSuppliers) {
+        return testCaseSuppliers.stream().map(supplier -> new TestCaseSupplier(supplier.name(), supplier.types(), () -> {
+            var testCase = supplier.supplier().get();
+
+            var newData = testCase.getData().stream().map(typedData -> {
+                if (typedData.data() instanceof BytesRef bytesRef) {
+                    var offset = randomIntBetween(0, 10);
+                    var extraLength = randomIntBetween(0, 10);
+                    var newBytesArray = randomByteArrayOfLength(bytesRef.length + offset + extraLength);
```
Comment (ivancea) [inline x-pack/plugin/esql/src/test/java/org/elasticsearch/xpack/esql/expression/function/AbstractFunctionTestCase.java:656]: Added extra length and randomization, as commented here: https://github.com/elastic/elasticsearch/pull/109205#discussion_r1620555420

File: x-pack/plugin/esql/src/test/java/org/elasticsearch/xpack/esql/expression/function/AbstractFunctionTestCase.java [hunk 2]
```diff
@@ -642,6 +642,46 @@ public interface ExpectedEvaluatorToString {
         Matcher<String> evaluatorToString(int nullPosition, TestCaseSupplier.TypedData nullData, Matcher<String> original);
     }
 
+    /**
+     * Modifies suppliers to generate BytesRefs with random offsets.
+     */
+    protected static List<TestCaseSupplier> randomizeBytesRefsOffset(List<TestCaseSupplier> testCaseSuppliers) {
```
Comment (nik9000) [inline x-pack/plugin/esql/src/test/java/org/elasticsearch/xpack/esql/expression/function/AbstractFunctionTestCase.java:648]: Could you try moving this into the ctor of `TypedData`? That'd get it everywhere for "free" which seems worth it.

Comment (ivancea) [reply x-pack/plugin/esql/src/test/java/org/elasticsearch/xpack/esql/expression/function/AbstractFunctionTestCase.java:648]: Copying from the original PR:

About the TypedData constructor, I would avoid it, as I could imagine some test manually adding offsets themselves for some usecase (?), as well as making things too complex (Now a plain constructor would also modify your data).
Not sure, but I feel like there's more control this way

Added:
I was thinking that would be interesting to replace the long chaing of supplier modifiers (failing types, nulls, etc) into a simpler "recommendedChain()". With that, we would easily apply any check everywhere. And for tests that don't want something, they can always do it the old way

**Review by nik9000 — APPROVED** [review]
Ok. Let's get it in and then build the recommendedChain thingy.

**Pull request merged**